hibernate-core/src/main/java/org/hibernate/dialect/temptable/TemporaryTable.java

```diff
@@ -198,61 +202,98 @@ public static TemporaryTable createIdTable(
-
-									if ( pluralAttribute.getSeparateCollectionTable() != null ) {
-										// Ensure that the FK target columns are available
-										ForeignKeyDescriptor keyDescriptor = pluralAttribute.getKeyDescriptor();
-										if ( keyDescriptor == null ) {
-											// This is expected to happen when processing a
-											// PostInitCallbackEntry because the callbacks
-											// are not ordered. The exception is caught in
-											// MappingModelCreationProcess.executePostInitCallbacks()
-											// and the callback is re-queued.
-											throw new IllegalStateException( "Not yet ready: " + pluralAttribute );
-										}
-										final ModelPart fkTarget = keyDescriptor.getTargetPart();
-										if ( !fkTarget.isEntityIdentifierMapping() ) {
-											final Value value = entityBinding.getSubclassProperty( pluralAttribute.getAttributeName() )
-													.getValue();
-											final Iterator<Selectable> columnIterator =
-													( (Collection) value ).getKey().getColumnIterator();
-											fkTarget.forEachSelectable(
-													(columnIndex, selection) -> {
-														final Selectable selectable = columnIterator.next();
-														if ( selectable instanceof Column ) {
-															final Column column = (Column) selectable;
-															columns.add(
-																	new TemporaryTableColumn(
-																			temporaryTable,
-																			column.getText( dialect ),
-																			selection.getJdbcMapping(),
-																			column.getSqlType(
-																					runtimeModelCreationContext.getMetadata()
-																			),
-																			column.getColumnSize(
-																					dialect,
-																					runtimeModelCreationContext.getMetadata()
-																			),
-																			column.isNullable()
-																	)
-															);
-														}
-													}
-											);
+					visitPluralAttributes( entityDescriptor, (pluralAttribute, attributeName) -> {
+						if ( pluralAttribute.getSeparateCollectionTable() != null ) {
+							// Ensure that the FK target columns are available
+							final ForeignKeyDescriptor keyDescriptor = pluralAttribute.getKeyDescriptor();
+							if ( keyDescriptor == null ) {
+								// This is expected to happen when processing a
+								// PostInitCallbackEntry because the callbacks
+								// are not ordered. The exception is caught in
+								// MappingModelCreationProcess.executePostInitCallbacks()
+								// and the callback is re-queued.
+								throw new IllegalStateException( "Not yet ready: " + pluralAttribute );
+							}
+							final ModelPart fkTarget = keyDescriptor.getTargetPart();
+							if ( !fkTarget.isEntityIdentifierMapping() ) {
+								final PersistentClass declaringClass = runtimeModelCreationContext.getBootModel()
+										.getEntityBinding( ( (EntityMappingType) pluralAttribute.getDeclaringType() ).getEntityName() );
```

Comment:
I think this will blow up with a class cast exception if you write a test for it. Can you check if this line is even hit for your test? Either way, you should be able to fix it by calling `findContainingEntityMapping()` instead of `getDeclaringType()`

Reply:
I did add a mapping with a (nested) embeddable which declares a plural attribute and I can confirm the line is indeed hit and its declaring type is still an EntityMappingType. Looking at the code looks like when creating `PluralAttributeMappingImpl`s we always pass `EntityPersister`s as declaring types, even if the `ManagedMappingType` would allow embeddables here.

Anyway `findContainingEntityMapping()` is safer, so I've updated the PR with that.

Reply:
>Looking at the code looks like when creating PluralAttributeMappingImpls we always pass EntityPersisters as declaring types, even if the ManagedMappingType would allow embeddables here.

This sounds problematic to me 😄 
Would you mind trying to pass the right thing there and see if it works?

Reply:
I'm trying but I see some breakage, I'm afraid we assumed to have an `EntityPersister` as declaring type in some places throughout the code and I'll need to review those if we wanna change that.

Reply:
> I'm afraid we assumed to have an EntityPersister as declaring type in some places throughout the code

So it looks like apart from some issues when creating the plural attribute mapping itself, which should be easily solvable, the only other thing that breaks is `org.hibernate.sql.results.graph.Initializer#isAttributeAssignableToConcreteDescriptor`. 

I don't know if we have a nice way of getting the containing entity type from an `AttributeMapping`, but if we don't maybe I could add something like `getDeclaringEntity()` which is only different to `getDeclaringType()` for plural attributes? Let me know if you have any tips, thanks.

Reply:
We already have `findContainingEntityMapping()` on `ManagedType`, so you could use `getDeclaringType().findContainingEntityMapping()` already.

Reply:
Sorry I keep forgetting about that method for some reason :disappointed: 

Looks like it should work, I'll push a separate commit with this change (also for `ToOneAttributeMapping`).